ml-agents/mlagents/trainers/trainer.py

```diff
@@ -165,97 +189,129 @@ def export_model(self, name_behavior_id: str) -> None:
-        """
-        raise UnityTrainerException("The end_episode method was not implemented.")
+        self._maybe_write_summary(self.get_step + len(trajectory.steps))
+        self._increment_step(len(trajectory.steps), trajectory.behavior_id)
 
-    def is_ready_update(self):
+    def _maybe_write_summary(self, step_after_process: int) -> None:
         """
-        Returns whether or not the trainer has enough elements to run update model
-        :return: A boolean corresponding to wether or not update_model() can be run
+        If processing the trajectory will make the step exceed the next summary write,
+        write the summary. This logic ensures summaries are written on the update step and not in between.
+        :param step_after_process: the step count after processing the next trajectory.
         """
-        raise UnityTrainerException("The is_ready_update method was not implemented.")
+        if step_after_process >= self.next_update_step and self.get_step != 0:
+            self._write_summary(self.next_update_step)
 
-    def update_policy(self):
+    @abc.abstractmethod
+    def end_episode(self):
         """
-        Uses demonstration_buffer to update model.
+        A signal that the Episode has ended. The buffer must be reset.
+        Get only called when the academy resets.
         """
-        raise UnityTrainerException("The update_model method was not implemented.")
+        pass
 
+    @abc.abstractmethod
     def create_policy(self, brain_parameters: BrainParameters) -> TFPolicy:
         """
         Creates policy
         """
-        raise UnityTrainerException("The create_policy method was not implemented.")
+        return None
 
+    @abc.abstractmethod
     def add_policy(self, name_behavior_id: str, policy: TFPolicy) -> None:
         """
         Adds policy to trainer
         """
-        raise UnityTrainerException("The add_policy method was not implemented")
+        pass
 
+    @abc.abstractmethod
     def get_policy(self, name_behavior_id: str) -> TFPolicy:
         """
         Gets policy from trainer
         """
-        raise UnityTrainerException("The get_policy method was not implemented.")
+        return None
 
-    def advance(self) -> None:
+    @abc.abstractmethod
+    def _is_ready_update(self):
+        """
+        Returns whether or not the trainer has enough elements to run update model
+        :return: A boolean corresponding to wether or not update_model() can be run
+        """
+        return False
+
+    @abc.abstractmethod
+    def _update_policy(self):
+        """
+        Uses demonstration_buffer to update model.
+        """
         pass
+
+    def advance(self) -> None:
+        """
+        Steps the trainer, taking in trajectories and updates if ready.
+        """
+        with hierarchical_timer("process_trajectory"):
+            for traj_queue in self.trajectory_queues:
+                if not traj_queue.empty():
```

Comment:
I think if there are ever multiple consumers here, you could run into problems - if there's one element in the queue, both could think it's non-empty, and one would try to grab it and the other would fail and raise an exception. Might be better to just to a try/except around the `get_nowait()` call, and avoid the `not empty()` check.

Reply:
Added an AgentManagerQueue.Empty exception and now using try/except here and in TC